### PR TITLE
[SPARK-13509][SPARK-13507][SQL] Support for writing CSV with a single function call

### PR DESCRIPTION
**Pull request by HyukjinKwon**
https://issues.apache.org/jira/browse/SPARK-13507
https://issues.apache.org/jira/browse/SPARK-13509
## What changes were proposed in this pull request?

This PR adds the support to write CSV data directly by a single call to the given path.

Several unitests were added for each functionality.
## How was this patch tested?

This was tested with unittests and with `dev/run_tests` for coding style

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@rxin I opened this PR because it looks writing `csv()` should be added anyway.
If I got the documentation stuff wrong, I will move that back.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -464,6 +464,12 @@ final class DataFrameWriter private[sql](df: DataFrame) {
    *   format("parquet").save(path)
    * }}}
    *
+   * You can set the following JSON-specific options for writing JSON files:
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This looks like it's in the wrong place?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -492,10 +498,33 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
+   *
+   * You can set the following CSV-specific options for writing CSV files:
+   * <li>`compression` or `codec` (default `null`): compression codec to use when saving to file.
+   * This should be the fully qualified name of a class implementing
+   * [[org.apache.hadoop.io.compress.CompressionCodec]] or one of the known case-insensitive
+   * shorten names(`bzip2`, `gzip`, `lz4`, and `snappy`). </li>
+   *
+   * @since 2.0.0
+   */
+  def csv(path: String): Unit = format("csv").save(path)
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This isn't a doc change; was it intentional?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -453,6 +453,12 @@ final class DataFrameWriter private[sql](df: DataFrame) {
    *   format("json").save(path)
    * }}}
    *
+   * You can set the following JSON-specific options for writing JSON files:
+   * <li>`compression` or `codec` (default `null`): compression codec to use when saving to file.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: just say compression, and don't mention codec.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: actually i'd remove codec support from the underlying source code, and only keep it for csv as an undocumented option for backward compatibility.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 4]
```diff
@@ -453,6 +453,12 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
+   * This should be the fully qualified name of a class implementing
+   * [[org.apache.hadoop.io.compress.CompressionCodec]] or one of the known case-insensitive
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I'd remove the mention of org.apache.hadoop.io.compress.CompressionCodec

The hadoop API is so difficult to use that I don't think it deserves to be user-facing here.

**Comment by HyukjinKwon** [comment]
@rxin Actually, do you think we need the `compression` option for Parquet and ORC as well (although I am not going to deal with them in this PR even if we need them)?

**Comment by rxin** [comment]
It'd be great to fix in a future pr.

for this one, let's also fix python?

**Comment by HyukjinKwon** [comment]
@rxin Sure.

**Comment by rxin** [comment]
LGTM

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Hm... It looks a bit weird. `ParquetHadoopFsRelationSuite`  the tests about types ("test all data types - ...") keep failing. This is also happening at other PRs I made.

I thought I wanted to submit a hot-fix but I found it actually works okay in my local.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@yhuai Could I ask that you have any clue on this? I think this is related with whole-code generation. This is happening some builds such as https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52154/consoleFull, https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52151/consoleFull and https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52153/consoleFull.

I could not find any clues until now.

**Comment by HyukjinKwon** [comment]
As this passes sometimes (e.g. https://github.com/apache/spark/pull/11016), I well restart.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
test this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I see that's a problem in new vecterizedreader. I missed the exception message. Looking deeper.

**Comment by HyukjinKwon** [comment]
@rxin BTW would you merge this if it looks good?

**Comment by rxin** [comment]
Thanks - merging this in master.

**Pull request closed without merging**